File: src/TelecommsSimulation.Core.CS/PhoneWithBlacklist.cs
using System.Threading.Tasks;

using TelecommsSimulation.Utilities;

namespace TelecommsSimulation.Core;

public class PhoneWithBlacklist : Phone
{
    public ConcurrentHashSet<string> Blacklist { get; }

    public PhoneWithBlacklist(string number, Coordinates location, double batteryLevel = 100) : base(number, location, batteryLevel) =>
        Blacklist = new ConcurrentHashSet<string>();

    public PhoneWithBlacklist(string number, Coordinates location, ConcurrentHashSet<string> blacklist, double batteryLevel = 100) : base(
        number, location, batteryLevel) =>
        Blacklist = blacklist;

    public override Task<CallAnswer> ReceiveCall(string from)
    {
        if (State is not PhoneState.Connected)
            throw new InvalidPhoneStateException();

        if (CallState is not CallState.Idle)
            throw new InvalidCallStateException();

        return Blacklist.Contains(from) ? Task.FromResult(CallAnswer.Declined) : base.ReceiveCall(from);
    }
}

File: src/TelecommsSimulation.Core/PhoneWithBlacklist.cpp
#include "pch.h"

#include "PhoneWithBlacklist.h"

#include "BaseStation.h"
#include "InvalidCallStateException.h"
#include "InvalidPhoneStateException.h"

using namespace System;
using namespace System::Threading::Tasks;

using namespace TelecommsSimulation::Utilities;

namespace TelecommsSimulation::Core
{
    ConcurrentHashSet<String^>^ PhoneWithBlacklist::Blacklist::get()
    {
        return _blacklist;
    }

    void PhoneWithBlacklist::Blacklist::set(ConcurrentHashSet<String^>^ value)
    {
        SetField<ConcurrentHashSet<String^>^>(_blacklist, value, "Blacklist");
    }

    PhoneWithBlacklist::PhoneWithBlacklist(String^ number, Coordinates location) : PhoneWithBlacklist(
        number, location, 100) {}

    PhoneWithBlacklist::PhoneWithBlacklist(String^ number, Coordinates location, double batteryLevel) : Phone(
        number, location, batteryLevel)
    {
        Blacklist = gcnew ConcurrentHashSet<String^>();
    }

    PhoneWithBlacklist::PhoneWithBlacklist(String^ number, Coordinates location,
                                           ConcurrentHashSet<String^>^ blacklist) :
        PhoneWithBlacklist(
            number, location, blacklist, 100) {}

    PhoneWithBlacklist::PhoneWithBlacklist(String^ number, Coordinates location,
                                           ConcurrentHashSet<String^>^ blacklist,
                                           double batteryLevel) : Phone(number, location, batteryLevel)
    {
        Blacklist = blacklist;
    }

    Task<CallAnswer>^ PhoneWithBlacklist::ReceiveCall(String^ from)
    {
        if (State != PhoneState::Connected)
            throw gcnew InvalidPhoneStateException();

        if (CallState != Core::CallState::Idle)
            throw gcnew InvalidCallStateException();

        if (Blacklist->Contains(from))
        {
            CallAnswered(this, gcnew CallAnsweredEventArgs(CallAnswer::Declined, Number, from));
            return Task::FromResult(CallAnswer::Declined);
        }

        return Phone::ReceiveCall(from);
    }
}
